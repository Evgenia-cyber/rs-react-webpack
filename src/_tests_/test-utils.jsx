/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import { render as rtlRender } from '@testing-library/react';

// Импортируем свои reducers
import homeReducer from '../redux/slices/homeSlice';
import detailsReducer from '../redux/slices/detailsSlice';

function render(
  ui,
  {
    preloadedState,
    store = configureStore({
      reducer: {
        home: homeReducer,
        details: detailsReducer,
      },
      preloadedState,
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { render };
