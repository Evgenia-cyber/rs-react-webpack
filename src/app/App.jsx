import React from 'react';
import Card from './components/Card';
// import img from "./img/img.jpg";

import './style.scss';

const App = () => {
  // const unused = 2;
  return (
    <>
      <h1 className="container">App</h1>
      <Card />
      <img src="./img/img.jpg" />
    </>
  );
};

export default App;
