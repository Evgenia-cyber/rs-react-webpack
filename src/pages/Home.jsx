import React from 'react';
import Card from '../components/Card';
import Form from '../components/Form';
import Paginator from '../components/Paginator';
import { minPageSize, sorts } from '../constants/constants';

const Home = () => {
  const [searchValue, setSearchValue] = React.useState('');

  const [sortBy, setSortBy] = React.useState(sorts[0]);

  const [pageSize, setPageSize] = React.useState(minPageSize);

  const [totalPages, setTotalPages] = React.useState('0');

  const [isLoading, setIsLoading] = React.useState(false);

  const [cards, setCards] = React.useState([]);

  return (
    <>
      <Form
        searchValue={searchValue}
        sortBy={sortBy}
        pageSize={pageSize}
        isLoading={isLoading}
        setCards={setCards}
        setIsLoading={setIsLoading}
        setSearchValue={setSearchValue}
        setSortBy={setSortBy}
        setPageSize={setPageSize}
        setTotalPages={setTotalPages}
      />
      {typeof cards !== 'string' && cards.length > 0 && (
        <>
          <Paginator
            searchValue={searchValue}
            sortBy={sortBy}
            pageSize={pageSize}
            totalPages={totalPages}
            setCards={setCards}
            setIsLoading={setIsLoading}
          />
          <div className="cards">
            {cards.map((card) => (
              <Card
                key={card.url}
                author={card.author}
                title={card.title}
                description={card.description}
                publishedAt={card.publishedAt}
                source={card.source}
                url={card.url}
                urlToImage={card.urlToImage}
              />
            ))}
          </div>
        </>
      )}
      {typeof cards === 'string' && <div className="no-results">{cards}</div>}
    </>
  );
};

export default Home;
