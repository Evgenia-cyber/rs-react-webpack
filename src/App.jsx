import React from 'react';
import { Route, Switch } from 'react-router-dom';
import Home from './pages/Home';

const App = () => (
  <div className="app-container">
    <Switch>
      <Route exact path="/" component={Home} />
    </Switch>
  </div>
);

export default App;
