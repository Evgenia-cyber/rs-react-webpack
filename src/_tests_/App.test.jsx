import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { BrowserRouter, Router } from 'react-router-dom';

import '@testing-library/jest-dom';
import { cleanup, render, screen } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { act } from 'react-dom/test-utils';
import pretty from 'pretty';

import store from '../redux/store';
import App from '../App';
import Home from '../pages/Home';

let container = null;
beforeEach(() => {
  // подготавливаем DOM-элемент, куда будем рендерить
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  // подчищаем после завершения
  ReactDOM.unmountComponentAtNode(container);
  container.remove();
  container = null;
  cleanup(); // для очистки после использования методов из библиотеки @testing-library
});

describe('Render App', () => {
  it('should render App without crashing', () => {
    act(() => {
      ReactDOM.render(
        <Provider store={store}>
          <BrowserRouter>
            <App />
          </BrowserRouter>
        </Provider>,
        container
      );
    });
  });
});

describe('Routing in the application', () => {
  it('should contain in url "/" immediately after starting the application', () => {
    const history = createMemoryHistory();
    render(
      <Provider store={store}>
        <Router history={history}>
          <App />
        </Router>
      </Provider>
    );

    expect(history.location.pathname).toEqual('/');
  });

  it('should be in Home page immediately after starting the application', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );

    expect(screen.getByText(/Search/)).toBeInTheDocument();
  });

  it('should render Home page immediately after starting the application (with snapshot)', () => {
    act(() => {
      ReactDOM.render(
        <Provider store={store}>
          <Home />
        </Provider>,
        container
      );
    });

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it('should be in NotFound page if unknown url', () => {
    const history = createMemoryHistory();
    const unknownRoute = '/some-unknown-route';
    history.push(unknownRoute);
    render(
      <Provider store={store}>
        <Router history={history}>
          <App />
        </Router>
      </Provider>
    );

    expect(screen.getByText(/No such page exists/)).toBeInTheDocument();
  });
});
