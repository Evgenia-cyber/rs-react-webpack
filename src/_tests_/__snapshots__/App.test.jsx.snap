// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routing in the application should render Home page immediately after starting the application (with snapshot) 1`] = `
"<form class=\\"form\\"><label class=\\"search\\" for=\\"search\\">
    <div class=\\"search-icon\\"><img src=\\"test-file-stub\\" alt=\\"magnifier\\"></div><input class=\\"search-bar\\" name=\\"search\\" type=\\"text\\" required=\\"\\" minlength=\\"3\\" maxlength=\\"40\\" placeholder=\\"Search...\\" value=\\"\\">
  </label>
  <div>
    <p class=\\"label\\">Sort by:</p>
    <div class=\\"sorts\\"><label for=\\"popularity\\"><input class=\\"sort\\" name=\\"sort\\" id=\\"popularity\\" type=\\"radio\\" value=\\"popularity\\" checked=\\"\\">popularity</label><label for=\\"relevancy\\"><input class=\\"sort\\" name=\\"sort\\" id=\\"relevancy\\" type=\\"radio\\" value=\\"relevancy\\">relevancy</label><label for=\\"publishedAt\\"><input class=\\"sort\\" name=\\"sort\\" id=\\"publishedAt\\" type=\\"radio\\" value=\\"publishedAt\\">publishedAt</label></div>
  </div><label class=\\"label\\" for=\\"page-size\\">Number of results per page (from 2 to 10):<input class=\\"page-size-input\\" name=\\"page-size\\" type=\\"number\\" min=\\"2\\" max=\\"10\\" value=\\"2\\"></label><button class=\\"submit-btn\\" type=\\"submit\\">Search</button>
</form>"
`;
